L1Trigger/Phase2L1ParticleFlow/interface/taus/L1HPSPFTauEmulator.h

```diff
@@ -0,0 +1,426 @@
+        sum_pt += sigP.hwPt;
+        sum_eta += sigP.hwPt * sigP.hwEta;
+        sum_phi += sigP.hwPt * sigP.hwPhi;
+      }
+    }
+
+    pt_t div_pt = 1;
+    if (sum_pt == 0) {
+      div_pt = 1;
+    } else {
+      div_pt = sum_pt;
+    }
+
+    tau_pt = sum_pt;
+    tau_eta = sum_eta / div_pt;
+    tau_phi = sum_phi / div_pt;
+
+    if (tau_pt > l1ct::Scales::makePtFromFloat(20.) && int_abs(tau_eta) < etaCutoff && leadSet == true &&
+        isocone_pt < tau_pt) {
+      Tau tau;
+      tau.hwPt = tau_pt;
+      tau.hwEta = tau_eta;
+      tau.hwPhi = tau_phi;
+      return tau;
+    } else {
+      Tau tau;
+      tau.hwPt = 0.;
+      tau.hwEta = 0.;
+      tau.hwPhi = 0.;
+      return tau;
+    }
+  }
+
+  inline std::vector<Tau> emulateEvent(std::vector<Particle>& parts, std::vector<Particle>& jets, bool jEnable) {
+    using namespace l1HPSPFTauEmu;
+
+    std::vector<Particle> parts_copy;
+    parts_copy.resize(parts.size());
+    std::transform(parts.begin(), parts.end(), parts_copy.begin(), [](const Particle& part) { return part; });
+    //sorting by pt
+    std::sort(parts_copy.begin(), parts_copy.end(), [](Particle i, Particle j) { return (i.hwPt > j.hwPt); });
+
+    //sorting jets by pt
+    std::vector<Particle> jets_copy;
+    jets_copy.resize(jets.size());
+    std::transform(jets.begin(), jets.end(), jets_copy.begin(), [](const Particle& jet) { return jet; });
+    std::sort(jets_copy.begin(), jets_copy.end(), [](Particle i, Particle j) { return (i.hwPt > j.hwPt); });
+
+    std::vector<Tau> taus;
+    std::vector<Tau> cleaned_taus;
+    taus.reserve(20);
+    std::vector<Particle> preseed;
+    preseed.reserve(144);
+
+    //jet seeds reserve
+    //4 for now
+    int jets_index = 0;
+    int jets_max = jets_copy.size();
+
+    std::vector<Particle> jseeds;
+    jseeds.reserve(4);
+
+    int parts_index = 0;
+    int parts_max = parts_copy.size();
+    //first find the seeds
+    while (preseed.size() < 128 && parts_index != parts_max) {
+      Particle pSeed = parts_copy.at(parts_index);
+
+      if (pSeed.hwPt > l1ct::Scales::makePtFromFloat(5.) && is_charged(pSeed) && int_abs(pSeed.hwEta) < etaCutoff) {
+        preseed.push_back(pSeed);
+      }
+
+      parts_index++;
+    }
+
+    std::vector<Particle> seeds;
+    seeds.reserve(16);  //up to 16 track + 4 jet seeds right now
+    std::vector<int>::size_type pseed_index = 0;
+    while (seeds.size() < 16 && pseed_index < preseed.size()) {
+      seeds.push_back(preseed.at(pseed_index));
+      pseed_index++;
+    }
+
+    //With jets
+    if (jEnable) {
+      while (jseeds.size() < 4 && jets_index != jets_max) {
+        Particle jSeed = jets_copy.at(jets_index);
+
+        if (jSeed.hwPt > l1ct::Scales::makePtFromFloat(20.) && int_abs(jSeed.hwEta) < etaCutoff) {
+          jseeds.push_back(jSeed);
+        }
+        jets_index++;
+      }
+    }
+    for (std::vector<int>::size_type i = 0; i != seeds.size(); i++) {
+      Particle seed = seeds.at(i);
```

Comment:
it will be perfectly safe here to use and optimize
Particle seed = seeds[i];

Also, do you need a copy here?